feat: Implement blocking queue SPI extension and add documentation

mingri31164 opened the pull request:
## **Problem**

1. Hard-coded `ResizableCapacityLinkedBlockingQueue` type check in `ServerThreadPoolDynamicRefresh.changePoolInfo()`
2. Cannot support custom blocking queue implementations like rejection policy

## **Solution**

1. Replace hardcoded type check with capability-based detection
2. Integrate `CustomBlockingQueue` SPI into `BlockingQueueTypeEnum.createBlockingQueue()`
3. Add `BlockingQueueManager` as unified facade for queue operations
4. Extract queue handling logic to separate `handleQueueChanges()` method
5. Support custom queue type configuration (applied on client restart via reflection)
6. Apply default capacity (1024) for custom queues

## **Key Changes**

1. Remove hardcoded type check in `changePoolInfo()`:
   - Replace `Objects.equals(RESIZABLE_LINKED_BLOCKING_QUEUE.getType(), parameter.getQueueType())` with capability detection
   - Extract to new `handleQueueChanges()` method for better separation
   - Use `BlockingQueueManager.canChangeCapacity()` for dynamic detection

2. Add `BlockingQueueManager` unified facade:
   - Create queue by type ID (built-in + custom SPI)
   - Validate queue configuration for all types
   - Check capacity change capability dynamically
   - Adjust capacity at runtime
   - Recognize queue type for built-in and SPI queues

3. Integrate `CustomBlockingQueue` SPI:
   - Enhance `BlockingQueueTypeEnum.createBlockingQueue()` to load SPI custom queues
   - Add `customOrDefaultQueue()` helper for SPI lookup with fallback
   - Register SPI via `ServiceLoaderRegistry`

4. Add documentation:
   - `queue-custom.md` for SPI registration guide
   - `queue-info.md` for queue type lifecycle explanation

## **Test Coverage**

1. `BlockingQueueSpiTest`: 10 scenarios including SPI registration, capacity capability detection, and design consistency
2. `QueueSpiIntegrationTest`: 6 scenarios validating hardcoded restriction removal and end-to-end flow
3. `ConfigServiceQueueCapacityTest`: 11 scenarios for custom queue default capacity handling

---

mingri31164 committed:
Add BlockingQueueManager

mingri31164 committed:
Remove hardcoded queues

mingri31164 committed:
Add queue extension and switching test code

mingri31164 committed:
Add a custom queue documents

mingri31164 committed:
Update queue switching description

mingri31164 committed:
fix: upgrade Mockito to 4.11.0

mingri31164 committed:
optimize ThreadPoolRebuilder

mingri31164 committed:
Add custom SPI blocking queue type support

mingri31164 committed:
fix:  Enable inline mock maker to support static mocks

mingri31164 committed:
Apply default capacity for custom queue types

mingri31164 committed:
Remove ThreadPoolRebuilder runtime queue switching

Createsequence commented on infra/common/src/main/java/cn/hippo4j/common/executor/support/BlockingQueueManager.java:
```diff
@@ -0,0 +1,179 @@
+@Slf4j
+public class BlockingQueueManager {
+
+    static {
```
This is the same registration code as in the BlockingQueueTypeEnum static block, so just keep one place.

Createsequence commented on starters/threadpool/server/src/main/java/cn/hippo4j/springboot/starter/core/ServerThreadPoolDynamicRefresh.java:
```diff
@@ -143,4 +145,29 @@ private void changePoolInfo(ThreadPoolExecutor executor, ThreadPoolParameter par
+            if (success) {
+                log.info("Queue capacity changed to: {}", parameter.getCapacity());
+            } else {
+                log.warn("Failed to change queue capacity to: {}", parameter.getCapacity());
```
when printing logs here, it is best to print the threadPoolId as well

mingri31164 commented:
Hi @Createsequence ,

I have a question regarding the Mockito version upgrade in this PR.

Currently: Mockito 4.11.0 (for JDK 17 compatibility)
Previous: Mockito 3.12.4

Should we keep the upgrade to 4.11.0 for JDK 17 support, or should I revert it back to 3.12.4?

Thank you for your feedback!

magestacks commented:
> Hi @Createsequence ,
> 
> I have a question regarding the Mockito version upgrade in this PR.
> 
> Currently: Mockito 4.11.0 (for JDK 17 compatibility) Previous: Mockito 3.12.4
> 
> Should we keep the upgrade to 4.11.0 for JDK 17 support, or should I revert it back to 3.12.4?
> 
> Thank you for your feedback!

@mingri31164 If the current version is not affected by this PR, keep the original version. The principle of a PR is to remain focused on a single purpose.

mingri31164 committed:
Clarify responsibilities and eliminate redundancy in BlockingQueueManager

mingri31164 committed:
rollback mockito version to 3.12.4

Createsequence commented:
Hi @mingri31164 

I noticed you also modified the front-end code for the console, but after the server starts, the front-end page doesn't support switching custom blocking queues.

Please confirm that after packaging your front-end code, you replaced the original files under the path `threadpool/server/console/src/main/resources/static`.

mingri31164 committed:
Update frontend static files with custom queue support

mingri31164 commented:
Hi @Createsequence ,

Thank you for your feedback. I'd like to clarify the situation:

### Original Frontend Behavior

**The original code already supports editing queue types in the admin console**, including all built-in queue types (ArrayBlockingQueue, LinkedBlockingQueue, etc.). There was no restriction on changing queue types in the frontend.
<img width="1461" height="1309" alt="44a3e41758dc060ecfaff10f7f4e7a33" src="https://github.com/user-attachments/assets/17d2701b-e5c1-46f4-a0fa-7f0b03e5bc4e" />

### My Changes

I followed the same pattern and added support for **custom blocking queues**:
- Added option `{ key: 99, display_name: 'CustomBlockingQueue (SPI)' }` 
- Added an input field for specifying the custom queue type ID
- This mirrors the `CustomRejectedPolicy (SPI)` implementation
<img width="830" height="1158" alt="18b0300bbfc85666a33851d095d5a586" src="https://github.com/user-attachments/assets/ecf03e2b-b439-4e6d-8364-b530e3b599ba" />


### Queue Type Change Mechanism

Both built-in and custom queue type changes follow the same restart-based mechanism:
- ✅ Configuration saved to database when edited
- ❌ NOT pushed to running clients immediately  
- ✅ Takes effect only when client application restarts
- ✅ Safe - no risk of task loss

### Next Steps

If you think rebuilding and deploying the frontend is not necessary, I will revert the frontend changes to the previous version instead.

Please let me know if you'd like me to proceed. Thank you!

Createsequence commented:
> Hi @Createsequence ,
> 
> Thank you for your feedback. I'd like to clarify the situation:
> 
> ### Original Frontend Behavior
> **The original code already supports editing queue types in the admin console**, including all built-in queue types (ArrayBlockingQueue, LinkedBlockingQueue, etc.). There was no restriction on changing queue types in the frontend. <img alt="44a3e41758dc060ecfaff10f7f4e7a33" width="1461" height="1309" src="https://private-user-images.githubusercontent.com/150683291/507059716-17d2701b-e5c1-46f4-a0fa-7f0b03e5bc4e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE3NTQzODcsIm5iZiI6MTc2MTc1NDA4NywicGF0aCI6Ii8xNTA2ODMyOTEvNTA3MDU5NzE2LTE3ZDI3MDFiLWU1YzEtNDZmNC1hMGZhLTdmMGIwM2U1YmM0ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyOVQxNjA4MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZWZhNDIwNTA1YTA4N2U3OGYwN2M5ZWJhMWNhOTEwN2UwZjE1Mzk0MGE4OWJmYTFhYzU3ZjUzNDhlMzJhODk5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jaauV8TEvT6k6V7cJysrjDgbFWABaPNwRyS32ega-yk">
> 
> ### My Changes
> I followed the same pattern and added support for **custom blocking queues**:
> 
> * Added option `{ key: 99, display_name: 'CustomBlockingQueue (SPI)' }`
> * Added an input field for specifying the custom queue type ID
> * This mirrors the `CustomRejectedPolicy (SPI)` implementation
> 
> <img alt="18b0300bbfc85666a33851d095d5a586" width="830" height="1158" src="https://private-user-images.githubusercontent.com/150683291/507062736-ecf03e2b-b439-4e6d-8364-b530e3b599ba.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE3NTQzODcsIm5iZiI6MTc2MTc1NDA4NywicGF0aCI6Ii8xNTA2ODMyOTEvNTA3MDYyNzM2LWVjZjAzZTJiLWI0MzktNGU2ZC04MzY0LWI1MzBlM2I1OTliYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyOVQxNjA4MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05YjJlZjU1NjBjNDdlOTM4MzA0OTQwYTM1MjllOWRlYmU1MWE3N2NmNDVmMzE0NDNhNDc5NzI5MDBjYmE5ODQyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.nqeFJaiohSwR2uhhYVipteHbuWM09RGv1yjhdK2HPCI">
> ### Queue Type Change Mechanism
> Both built-in and custom queue type changes follow the same restart-based mechanism:
> 
> * ✅ Configuration saved to database when edited
> * ❌ NOT pushed to running clients immediately
> * ✅ Takes effect only when client application restarts
> * ✅ Safe - no risk of task loss
> 
> ### Next Steps
> If you think rebuilding and deploying the frontend is not necessary, I will revert the frontend changes to the previous version instead.
> 
> Please let me know if you'd like me to proceed. Thank you!

Your solution is undoubtedly reasonable, and I don't think there's anything wrong with it.
Just considering that the front-end and back-end code are in the same project, I think the front-end and back-end code under the same theme should be submitted together, just like you did in [this commit](https://github.com/opengoofy/hippo4j/pull/1612/commits/8c670ffe37fec3209009354c51532d6d8f779c75) is done like that.
Anyway, this is a great PR with clean code and detailed comments. Once we finish the final review, we will merge it as soon as possible :)

mingri31164 committed:
Allow passing queue capacity to custom blocking queue SPI

Createsequence reviewed (approved):
good job
